keps/sig-scheduling/3280-guarantee-pdb-when-preemption-happens/README.md

```diff
@@ -0,0 +1,981 @@
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+This design proposal suggests adding a field `DisruptionPolicy` in the `PriorityClass` API to indicate 
+whether or not make `PodDisruptionBudget` guaranteed during the scheduler preemption process, in order to meet the 
+needs of users who do not want PDB violations to occur during scheduler preemption due to some reasons such as 
+high availability of services(https://github.com/kubernetes/kubernetes/issues/91492#issuecomment-1029484252). 
+At the same time, to prevent situations where core components or necessary add-ons cannot be scheduled due to the 
+inability to violate PDBs, priority ClassName of `system-cluster-critical` or `system-node-critical` will have the 
+exemption and not be required to comply with this configuration, and the behavior will be consistent with the current one.
```

Comment:
Hmm - I'm somewhat on the fence with this feature. Without this feature, if I was running higher-prioirty pods I was guaranteed that my pods will get scheduled, because they can preempt something else.

Now, having a higher-priority pods doesn't buy me much, because if the cluster is full (and there is no place to schedule anything else) and lower-level PDB targets are disrupted to their levels, I'm stuck.

I'm not entirely sure I'm buying the usecase for this feature - if my pods can't tolerate disruption caused by some other pods, are they really of lower-priority than the others?

I think this change would introduce enormous confusion to users - I'm higher-priority, but I'm still not able to run in the cluster even though lower-priority stuff is running there.

This feature effectively means that at least `MinAvailable` replicas from each set of pods protected by PDB in those priority class effectively isn't in this priority class, but rather is in the highest-possible priority class (except from system-cluster-critical and system-node-critical).

That is opening the system for gaming the system:
- I'm running with the lowest priority-class, so let me mark this bit and set PDB target high enough and I will be guaranteed to run a huge number of my pods.


This makes this feature really risky imho. I would really like us to spend more time on figuring out other options we can address the main usecase for it.

Reply:
> if my pods can't tolerate disruption caused by some other pods, are they really of lower-priority than the others?

Somehow this represents a semantics of "scheduled in a low-priroity, but once running it's a high-priority that can only bare minimum disruption". Stateful workloads that have associated (local) storage are of this type.

> I'm higher-priority, but I'm still not able to run in the cluster even though lower-priority stuff is running there.

Yes, it sounds the case. But in practice, only very restricted PriorityClasses are created in this way, under cluster SRE's control. In other words, priority value is not the only factor to impact whether a pod can preempt others.

> That is opening the system for gaming the system:

So when opt-in this feature, SRE admin should be cautious when creating this kind of PriorityClass, they should know what they're doing.

--

Providing a data point about gaming: I'd say the current priority/preemption system _can_ be gamed already:

- a high-priority pod H requests 2 CPUs, without PDB, scheduled on node A
- a low-priority pod L requests 2 CPUs, with PDB protected, scheduled on node B
- node A & B are both fully utilized
- now it comes in a system-critical pod S requesting 2CPUs. In this case, although pod H holds a higher priority than L, given scheduler honors PDB in a best-efforts manner, H gets preempted instead of L

Believe it or not, this is one reason some end-users don't use PDB in their prod env. So a side-product of this KEP is to give an option for SRE admin to disable this behavior to avoid abuse, although it will be still enabled by default for backwards compatibility.

Reply:
Not sure I understand the problem. H should have a PDB if can't tolerate disruption. But if it's disrupted, the deployment would create another pod that should conquer over L.

> So a side-product of this KEP is to give an option for SRE admin to disable this behavior to avoid abuse.

How? Are you proposing a mechanism to do preemption without looking at PDBs?

Reply:
> How? Are you proposing a mechanism to do preemption without looking at PDBs?

Yes, behave like one type of disruption policy. But I don't want to derail the topic here, where we should focus on the general idea of disruption policy, as well as different controls on PDB.

> the deployment would create another pod that should conquer over L.

There can be medium priorityclass in between, also the combination of PDB violated/not-violated on a node can be pretty complex. Take a step back, even if we _could_ prove eventually L will fail on this abuse game, it's still nonsense to me if it's based on the cost of so many extra steps (casdaded premeption, re-create pods).

Reply:
> Hmm - I'm somewhat on the fence with this feature. Without this feature, if I was running higher priority pods I was guaranteed that my pods will get scheduled, because they can preempt something else.

Maybe not. The value in priority class is not the only criteria that determines whether a Pod can be preempted. There is also another field called preemptionPolicy in PriorityClass that sets the preemption strategy. A PriorityClass with a high Value may not be able to preempt other lower priority Pods. https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/#non-preempting-priority-class

> I'm not entirely sure I'm buying the use case for this feature - if my pods can't tolerate disruption caused by some other pods, are they really of lower-priority than the others?

It's a common occurrence in real production environments that some services or jobs have a lower priority and can be preempted, but once they reach the `MinAvailable`, every pod‘s importance increases and they can no longer be preempted arbitrarily. In addition to Stateful workloads as mentioned by Wei, there are also distributed training jobs for AI and BigData, which can run and be preempted by other high priority pods, but if they fall below MinAvailable, the entire task will fail directly, leading to a bigger impact.

> That is opening the system for gaming the system: I'm running with the lowest priority-class, so let me mark this bit and set PDB target high enough and I will be guaranteed to run a huge number of my pods.

Yes, this is why we placed the field in PriorityClass instead of PDB based on @liggitt 's suggestion. https://github.com/kubernetes/enhancements/pull/3755#discussion_r1083853660 . Only cluster administrators can create and configure PriorityClass, and they can decide which PriorityClass needs to be protected, avoiding any gaming behavior among regular users.

In addition, the objective of this design document is to clarify the behavior of PDBs during scheduling. In the past, users were unable to determine if their services would be preempted by the scheduler under PDBs. To ensure the integrity of service PDBs, simply increasing the value of PriorityClass is not effective and will only lead to low-priority tasks becoming high priority, thereby impacting the scheduling of pods with high priority.

Reply:
> Maybe not. The value in priority class is not the only criteria that determines whether a Pod can be preempted. There is also another field called preemptionPolicy in PriorityClass that sets the preemption strategy. A PriorityClass with a high Value may not be able to preempt other lower priority Pods. https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/#non-preempting-priority-class

I don't agree with this one. This is saying that "my pods can't preempt someone else" - I own the decision that my pods can't even preempt lower-priority pods. This is the right permission to have, as I don't affect higher-priority pods (I'm not saying you can't preempt me even if you're more important).


> Yes, this is why we placed the field in PriorityClass instead of PDB based on @liggitt 's suggestion. https://github.com/kubernetes/enhancements/pull/3755#discussion_r1083853660 . Only cluster administrators can create and configure PriorityClass, and they can decide which PriorityClass needs to be protected, avoiding any gaming behavior among regular users.

I spent some time thinking about it and I'm less sceptical about it now.
I would actually move what you have under "Notes/Constraints/Caveats (Optional)" section to "Risks and Mitigations" to explicitly describe the problem (low-prio pods becoming high-prio pods) and an explicit mitigation we have for that (giving this power only to administrator).